apps/dgst.c

```diff
@@ -405,13 +406,8 @@ int dgst_main(int argc, char **argv)
     } else {
         const char *sig_name = NULL;
         if (!out_bin) {
-            if (sigkey != NULL) {
-                const EVP_PKEY_ASN1_METHOD *ameth;
-                ameth = EVP_PKEY_get0_asn1(sigkey);
-                if (ameth)
-                    EVP_PKEY_asn1_get0_info(NULL, NULL,
-                                            NULL, NULL, &sig_name, ameth);
-            }
+            if (sigkey != NULL)
+                sig_name = EVP_PKEY_get0_first_alg_name(sigkey);
```

Comment:
This is very dodgy.  One alias to the HMAC signature algorithm and the test may randomly go **BAMF**, because you might end up with, oh I don't know, "id-Hmac".  This is a case where the application has more accurate knowledge than libcrypto, as it really has *all* the data it needs to know exactly what names it's asking for.  It's quite a simple program after all...

Reply:
Err, no. We're using it for HMAC here, but in theory you can sign with any type of key - so we really have to get the name from the key.